druid-shell/src/platform/gtk/window.rs

```diff
@@ -900,16 +946,34 @@ impl WindowHandle {
         }
     }
 
-    pub fn open_file_sync(&mut self, options: FileDialogOptions) -> Option<FileInfo> {
-        self.file_dialog(FileDialogType::Open, options)
-            .ok()
-            .map(|s| FileInfo { path: s.into() })
+    pub fn open_file(&mut self, options: FileDialogOptions) -> Option<FileDialogToken> {
+        if let Some(state) = self.state.upgrade() {
+            let tok = FileDialogToken::new(state.next_token());
+            state.defer(DeferredOp::Open(options, tok));
+            Some(tok)
+        } else {
+            None
+        }
     }
 
-    pub fn save_as_sync(&mut self, options: FileDialogOptions) -> Option<FileInfo> {
-        self.file_dialog(FileDialogType::Save, options)
-            .ok()
-            .map(|s| FileInfo { path: s.into() })
+    pub fn save_as(&mut self, options: FileDialogOptions) -> Option<FileDialogToken> {
+        if let Some(state) = self.state.upgrade() {
+            let tok = FileDialogToken::new(state.next_token());
+            state.defer(DeferredOp::SaveAs(options, tok));
+            Some(tok)
+        } else {
+            None
+        }
+    }
+
+    pub fn save_as_sync(&mut self, _options: FileDialogOptions) -> Option<FileInfo> {
```

Comment:
am I understanding his correctly that with this patch, `save_as_sync` won't work on gtk, and `save_as` won't work elsewhere?

Reply:
Right. The current situation is that druid papers over the druid-shell API nastiness. I did a quick `rg` on runebender and it looks like it should still work, because AFAICT you're using the druid commands instead of manually calling the `WIndowHandle` methods?

The plan is to implement `save_as` and `open_file` for the other backends (just windows and mac, really, because X11 and web don't support file dialogs yet anyway; and I can do windows soon). Then we delete the `xxx_sync` methods on `WindowHandle` and remove the fallback hack in druid.

Any druid users that are using the `SHOW_OPEN_PANEL`, etc, commands shouldn't even notice the change (except that they'll stop getting dropped event warnings).

Reply:
Right. The current situation is that druid papers over the druid-shell API nastiness. I did a quick `rg` on runebender and it looks like it should still work, because AFAICT you're using the druid commands instead of manually calling the `WIndowHandle` methods?

The plan is to implement `save_as` and `open_file` for the other backends (just windows and mac, really, because X11 and web don't support file dialogs yet anyway; and I can do windows soon). Then we delete the `xxx_sync` methods on `WindowHandle` and remove the fallback hack in druid.

Any druid users that are using the `SHOW_OPEN_PANEL`, etc, commands shouldn't even notice the change (except that they'll stop getting dropped event warnings).